pylabnet/hardware/ZI_HDAWG/zi_hdawg.py

```diff
@@ -0,0 +1,475 @@
+                    permissible range {list(range(14))}."
+            )
+            return
+
+        self.hd.seti(f'awgs/{self.index}/time', sampling_rate_index)
+        self.hd.log.info(
+            f"AWG {self.index}: Changed sampling rate \
+                 to index {sampling_rate_index}."
+        )
+
+    def start(self):
+        """ Start AWG"""
+        self.module.set('awg/enable', 1)
+        self.hd.log.info(f"AWG {self.index}: Started.")
+
+    def stop(self):
+        """ Stop AWG"""
+        self.module.set('awg/enable', 0)
+        self.hd.log.info(f"AWG {self.index}: Stopped.")
+
+    def compile_upload_sequence(self, sequence):
+        """ Compile and upload AWG sequence to AWG Module
+
+        :awgModule: Instance of AWGModule class.
+        """
+
+        # First check if all values have been replaced in sequence:
+        if not sequence.is_ready():
+            self.log.error("Sequence is not ready:\
+                 Not all placeholders have been replaced.")
+            return
+
+        self.module.set('compiler/sourcestring', sequence.sequence)
+        # Note: when using an AWG program from a source file
+        # (and only then), the compiler needs to
+        # be started explicitly with awgModule.set('compiler/start', 1)
+        while self.module.getInt('compiler/status') == -1:
+            time.sleep(0.1)
+
+        if self.module.getInt('compiler/status') == 1:
+            # compilation failed, raise an exception
+            self.hd.log.warn(self.module.getString('compiler/statusstring'))
+
+        if self.module.getInt('compiler/status') == 0:
+            self.hd.log.info(
+                "Compilation successful with no warnings, \
+                     will upload the program to the instrument."
+            )
+        if self.module.getInt('compiler/status') == 2:
+            self.hd.log.warn(
+                "Compilation successful with warnings, \
+                    will upload the program to the instrument."
+            )
+            self.hd.log.warn(
+                f"Compiler warning: \
+                     {self.module.getString('compiler/statusstring')}"
+            )
+
+        # Wait for the waveform upload to finish
+        time.sleep(0.2)
+        i = 0
+        while (self.module.getDouble('progress') < 1.0) and (self.module.getInt('elf/status') != 1):
+            self.hd.log.info("{} progress: {:.2f}".format(i, self.module.getDouble('progress')))
+            time.sleep(0.2)
+            i += 1
+        self.hd.log.info(
+            "{} progress: {:.2f}".format(i, self.module.getDouble('progress'))
+        )
+        if self.module.getInt('elf/status') == 0:
+            self.hd.log.info("Upload to the instrument successful.")
+        if self.module.getInt('elf/status') == 1:
+            self.hd.log.warning("Upload to the instrument failed.")
+
+    def dyn_waveform_upload(self, index, waveform1, waveform2=None):
```

Comment:
I think the index will be the main form of addressing and keeping track of the waveform on the level of the AWG. It would be possible to write some sort of waveform-registry which keeps track of which waveform index is associated with which wave- we can think about this. Alternatively, we can store waveforms as `.csv` files, and name them accordingly.

Reply:
just pointing out that we absolutely need some experiment-specific labeling of waveforms at the top level. probably doesn't need to be at this level, I guess.